**Student Name:**  Camille Settles

**Student ID:**   @02986989

**Course:** MATH 014 – Introduction to Data Science  

**Project 2:**  Part 1: Data Wrangling & Descriptive Statistics 

## **Introduction & Context:**

This dataset contains performance and contextual information for **566** Chicago Public Schools (CPS) during the **2011–2012** academic year. The purpose of the dataset was to help parents, educators, and policymakers better understand school quality and support data-driven decisions at both the individual and district levels. The dataset offers a comprehensive view of school operations, student outcomes, and community factors that may influence educational success. 

It includes:
- **Academic standards** such as standardized test performance (ISAT, EXPLORE, PLAN, ACT), graduation rates, and college eligibility percentages.
- **Operational indicators** including school safety scores, student attendance, misconduct rates, and teacher attendance.
- **Community and environmental metrics** such as parent engagement levels, type of school (elementary, middle, or high), neighborhood ward, police district, and geographic location.  

This approach allows for a deeper understanding of school effectiveness beyond just test scores.



## **Dataset Source:**

**Title:** Chicago Public Schools – Progress Report Cards (2011–2012)  

**Source:** https://catalog.data.gov/dataset/chicago-public-schools-progress-report-cards-2011-2012

**File Name:** `Chicago_Public_Schools_-_Progress_Report_Cards__2011-2012_ (1).csv`


## Section I: Dataset Overview & Structure

This section focuses on understanding the overall shape and structure of the dataset. We will examine the dimensions, data types, and presence of missing values.

#### 1. What is the shape and structure of the dataset?

In [301]:
import pandas as pd

df = pd.read_csv("Chicago_Public_Schools_-_Progress_Report_Cards__2011-2012_ (1).csv") # Load the dataset

print("Shape of dataset:", df.shape) # Display the number of rows and columns

print("\n General Information of the dataset: \n", df.describe())  # Runs .describe() to display summary stats for all numerical columns

print("\n Info on Columns and Data Types:",df.info) # Display basic info on columns and data types


Shape of dataset: (566, 79)

 General Information of the dataset: 
            School ID      ZIP Code  Safety Score  Environment Score  \
count     566.000000    566.000000    513.000000         513.000000   
mean   609681.791519  60629.727915     49.504873          47.766082   
std      8831.060946     20.248024     20.110837          16.215584   
min    400018.000000  60605.000000      1.000000           1.000000   
25%    609873.250000  60618.000000     35.000000          37.000000   
50%    610052.500000  60625.000000     48.000000          47.000000   
75%    610212.750000  60639.000000     61.000000          58.000000   
max    610544.000000  60827.000000     99.000000          99.000000   

       Instruction Score  Average Student Attendance  \
count         513.000000                  565.000000   
mean           48.288499                   92.313805   
std            17.417176                    5.993053   
min             1.000000                   57.900000   
25%         

In [303]:
print("First 5 rows of the dataset:") # Preview the first 5 rows of the dataset
print(df.head())

First 5 rows of the dataset:
   School ID                             Name of School  \
0     609966        Charles G Hammond Elementary School   
1     610539            Marvin Camras Elementary School   
2     609852           Eliza Chappell Elementary School   
3     609835         Daniel R Cameron Elementary School   
4     610521  Sir Miles Davis Magnet Elementary Academy   

  Elementary, Middle, or High School          Street Address     City State  \
0                                 ES         2819 W 21st Pl   Chicago    IL   
1                                 ES       3000 N Mango Ave   Chicago    IL   
2                                 ES      2135 W Foster Ave   Chicago    IL   
3                                 ES  1234 N Monticello Ave   Chicago    IL   
4                                 ES      6730 S Paulina St   Chicago    IL   

   ZIP Code    Phone Number  \
0     60623  (773) 535-4580   
1     60634  (773) 534-2960   
2     60625  (773) 534-2390   
3     60651  (773

In [305]:
print("\nLast 5 rows of the dataset:") # Preview the last 5 rows of the dataset
print(df.tail())


Last 5 rows of the dataset:
     School ID                                    Name of School  \
561     610506      TEAM Englewood Community Academy High School   
562     610350           Robert A Black Magnet Elementary School   
563     610154                Martin A Ryerson Elementary School   
564     610384  Infinity Math Science and Technology High School   
565     609870                Daniel J Corkery Elementary School   

    Elementary, Middle, or High School       Street Address     City State  \
561                                 HS  6201 S Stewart Ave   Chicago    IL   
562                                 ES   9101 S Euclid Ave   Chicago    IL   
563                                 ES  646 N Lawndale Ave   Chicago    IL   
564                                 HS  3120 S Kostner Ave   Chicago    IL   
565                                 ES  2510 S Kildare Ave   Chicago    IL   

     ZIP Code    Phone Number  \
561     60621  (773) 535-3530   
562     60619  (773) 535-63

**Explanation:**

The dataset has **566 rows** and **79 columns**. Each row represents one school within the Chicago Public School (CPS) system, and each column provides a variable related to school performance, environment, location, or engagement.

- **The `.shape` method** confirms the number of rows and columns in the dataset. This helps ensure that the dataset loaded correctly and matches the structure described in the data source.

- **The `.info()` method** provides important structural metadata:
  - Lists each column name.
  - Shows the number of non-null entries.
  - Displays the data type (`object`, `int64`, or `float64`) for each column.

From this output:
- Most columns are `object` type, meaning they contain categorical or text-based data (e.g., *Name of School*, *City*, *Performance Status*).
- Some columns, like `Safety Score` and `Instruction Score`, are `float64`, indicating continuous numerical values.
- Example: The `Link` column has 565 non-null values (1 missing), highlighting a small gap in the dataset.

- **The `.head()` and `.tail()` methods** display the first and last five rows of the dataset, which looks at the data’s layout. This data confirms that:
  - Column headers are correctly labeled and appear as expected (e.g., *School ID*, *ZIP Code*).
  - Data is consistently formatted across rows.
  - Entries are aligned under the appropriate columns.

These tools describe the dataset’s structure and identify missing data or formatting concerns before cleaning and analysis.


####  2. Identify the Variable Types


In [309]:
df.dtypes.value_counts() # Counts how many columns fall under each data type


object     56
float64    15
int64       8
Name: count, dtype: int64

**Explanation:**

The code `df.dtypes.value_counts()` counts how many columns fall under each data type. It summarizes the structure of the data set and the variables. This method shows how many columns fall into each data type. 

The dataset contains:

- **56 `object` columns** — These include school names, categories, labels, and other non-numeric entries.

- **15 `float64` columns** — These are decimal numbers, typically used for percentage-based metrics (e.g., **Safety Score**, **Average Attendance**).

- **8 `int64` columns** — These are whole numbers, representing **IDs**, **ZIP codes**, or **count-based values**.


#### 3. Are there any missing values?

In [319]:
# Count and sort missing values in descending order
missing_values = df.isnull().sum()  # Calculates the total number of missing (`null`) entries for each column in the dataset.

missing_values[missing_values > 0].sort_values(ascending=False) # Filters out columns that have no missing values.
print(missing_values[missing_values > 0].sort_values(ascending=False))




ISAT Value Add Math           98
ISAT Value Add Read           98
ISAT Exceeding Math %         90
ISAT Exceeding Reading %      90
Safety Score                  53
Environment Score             53
Instruction Score             53
Link                           1
Average Student Attendance     1
dtype: int64


**Explanation:**

This output reveals that several columns contain missing values, primarily in **performance-related numerical features**. These missing values likely result from differences in reporting between schools.

The columns with the most missing values are:

- **ISAT Value Add Math** — 98 missing  
- **ISAT Value Add Read** — 98 missing  
- **ISAT Exceeding Math %** — 90 missing  
- **ISAT Exceeding Reading %** — 90 missing  
- **Safety Score** — 53 missing  
- **Environment Score** — 53 missing  
- **Instruction Score** — 53 missing  
- **Link** — 1 missing  
- **Average Student Attendance** — 1 missing  

These values are not unexpected in public school datasets, where data collection can vary from school to school.

**Plan for Handling Missing Data:**

- For **numerical columns** (e.g., test scores, attendance rates):  
  Use **mean or median imputation** to preserve the dataset while reducing bias.

- For **categorical or optional fields**:  
  Replace missing values with a placeholder like `"N/A"` if the column is not critical for analysis.


## Section II: Data Cleaning & Indexing 

This section focuses on preparing the dataset for analysis by removing redundant or irrelevant information and filtering data using logical conditions. It ensures that the dataset is clean, unique, and ready for further processing.

#### 4. Check for and Remove Duplicates

In [324]:
df.duplicated().sum() # Check if there are any duplicate rows
df[df.duplicated()] # Returns 0

Empty DataFrame
Columns: [School ID, Name of School, Elementary, Middle, or High School, Street Address, City, State, ZIP Code, Phone Number, Link , Network Manager, Collaborative Name, Adequate Yearly Progress Made? , Track Schedule, CPS Performance Policy Status, CPS Performance Policy Level, Healthy Schools Certified?, Safety Icon , Safety Score, Family Involvement Icon, Family Involvement Score, Environment Icon , Environment Score, Instruction Icon , Instruction Score, Leaders Icon , Leaders Score , Teachers Icon , Teachers Score, Parent Engagement Icon , Parent Engagement Score, Parent Environment Icon, Parent Environment Score, Average Student Attendance, Rate of Misconducts (per 100 students) , Average Teacher Attendance, Individualized Education Program Compliance Rate , Pk-2 Literacy %, Pk-2 Math %, Gr3-5 Grade Level Math %, Gr3-5 Grade Level Read % , Gr3-5 Keep Pace Read %, Gr3-5 Keep Pace Math %, Gr6-8 Grade Level Math %, Gr6-8 Grade Level Read %, Gr6-8 Keep Pace Math%, Gr6-8 Keep Pace Read %, Gr-8 Explore Math %, Gr-8 Explore Read %, ISAT Exceeding Math %, ISAT Exceeding Reading % , ISAT Value Add Math, ISAT Value Add Read, ISAT Value Add Color Math, ISAT Value Add Color Read, Students Taking  Algebra %, Students Passing  Algebra %, 9th Grade EXPLORE (2009) , 9th Grade EXPLORE (2010) , 10th Grade PLAN (2009) , 10th Grade PLAN (2010) , Net Change EXPLORE and PLAN, 11th Grade Average ACT (2011) , Net Change PLAN and ACT, College Eligibility %, Graduation Rate %, College Enrollment Rate %, College Enrollment (number of students) , General Services Route , Freshman on Track Rate %, RCDTS Code, X_COORDINATE, Y_COORDINATE, Latitude, Longitude, Community Area Number, Community Area Name, Ward, Police District, Location]
Index: []

[0 rows x 79 columns]

**Explanation:**

The code `.duplicated().sum()` checks for duplicate rows in the dataset and returns the total number. The output shows `0`, meaning there are **no duplicate rows**, and nothing needs to be removed. The code `df.drop_duplicates()` can be used to check again. 




In [327]:
df = df.drop_duplicates() # Remove duplicate rows if they exist


**Explanation:**

The code `df.drop_duplicates()` removes any duplicate rows if they exist. There is no output, meaning that there are no duplicates. 

#### 5. Provide Examples of DataFrame Filtering Using Logical Conditions



In [331]:
very_safe_schools = df[df["Safety Score"] > 90] # Filter Schools with Safety Score > 90 (Very Safe)
very_safe_schools[["Name of School", "Safety Score"]] # Displays name and safety score of filtered schools

Name of School  Safety Score
8        Walter Payton College Preparatory High School          98.0
11                   Abraham Lincoln Elementary School          99.0
17           Northside College Preparatory High School          99.0
26                       Edison Park Elementary School          95.0
33             Alexander Graham Bell Elementary School          99.0
49                       Oriole Park Elementary School          99.0
151   Mary E Courtenay Elementary Language Arts Center          99.0
203                   Ellen Mitchell Elementary School          99.0
210        Stephen Decatur Classical Elementary School          99.0
239         James E McDade Elementary Classical School          99.0
251                     Norwood Park Elementary School          99.0
277                   John J Audubon Elementary School          95.0
280        Edgar Allan Poe Elementary Classical School          99.0
285       Annie Keller Elementary Gifted Magnet School          99.0
319                   James G Blaine Elementary School          99.0
322                Augustus H Burley Elementary School          99.0
330                           Talman Elementary School          99.0
338            Hawthorne Elementary Scholastic Academy          92.0
373                 Whitney M Young Magnet High School          95.0
384      William Jones College Preparatory High School          92.0
466                        Edgebrook Elementary School          99.0
488              Northside Learning Center High School          99.0
505                LaSalle Elementary Language Academy          99.0
520                        Sauganash Elementary School          99.0
522                         Wildwood Elementary School          99.0
542  Thomas A Edison Regional Gifted Center Element...          91.0

**Explanation:**

This filter selects schools with a Safety Score greater than 90, identifying institutions considered very safe environments. This helps prioritize schools with a strong safety record. The output displays the names and safety scores of the safest schools. These schools stand out for their protective environments.

In [334]:
high_schools = df[df["Elementary, Middle, or High School"] == "HS"] # Filter Schools by Type (High Schools)
high_schools[["Name of School", "Elementary, Middle, or High School"]].head() # Shows first few results of filter

Name of School  \
8    Walter Payton College Preparatory High School   
15               Manley Career Academy High School   
17       Northside College Preparatory High School   
28  Michele Clark Academic Prep Magnet High School   
30                    Uplift Community High School   

   Elementary, Middle, or High School  
8                                  HS  
15                                 HS  
17                                 HS  
28                                 HS  
30                                 HS

**Explanation:**

This filter only selects high schools from the dataset using a categorical condition. This is useful when analyzing conditions specific to grade level. The output shows the first few high schools.

In [346]:
# Convert involvement columns to numeric 
df["Family Involvement Score"] = pd.to_numeric(df["Family Involvement Score"], errors='coerce') # Converts to numeric
df["Teachers Score"] = pd.to_numeric(df["Teachers Score"], errors='coerce') # Converts teacher scores to numeric format
df["Parent Engagement Score"] = pd.to_numeric(df["Parent Engagement Score"], errors='coerce') # Converts parent engagement scores to numeric format

# Filter schools with high Family, Teacher, and Parent involvement scores
involvement_scores = df[
    (df["Family Involvement Score"] > 60) & # Family score must be greater than 60
    (df["Teachers Score"] > 60) &  # Teacher score must be greater than 60
    (df["Parent Engagement Score"] > 60) # Parent Engagement score must be greater than 60
]

# Display selected columns for filtered schools
involvement_scores[["Name of School", "Family Involvement Score", "Teachers Score", "Parent Engagement Score"]].head()


Name of School  Family Involvement Score  \
277              John J Audubon Elementary School                      73.0   
285  Annie Keller Elementary Gifted Magnet School                      97.0   
365  Jacqueline B Vaughn Occupational High School                      72.0   
405             Frederick Stock Elementary School                      99.0   
488         Northside Learning Center High School                      99.0   

     Teachers Score  Parent Engagement Score  
277            65.0                     61.0  
285            82.0                     68.0  
365            69.0                     63.0  
405            81.0                     69.0  
488            84.0                     62.0

**Explanation:**

This filter selects schools where Family Involvement, Teachers Score, and Parent Engagement Score are all above 60. These values indicate strong collaboration among staff, families, and school communities. The output shows schools with strong involvement across all three areas. These schools may be strong models for school engagement strategies.



## Section III: Descriptive Statistics 

#### 6. Use .describe() on numerical columns:

In [351]:
df.describe() # Summary statistics (count, mean, std, min, max, etc.) for all numerical columns in the dataset


School ID      ZIP Code  Safety Score  Family Involvement Score  \
count     566.000000    566.000000    513.000000                297.000000   
mean   609681.791519  60629.727915     49.504873                 50.602694   
std      8831.060946     20.248024     20.110837                 18.549160   
min    400018.000000  60605.000000      1.000000                  6.000000   
25%    609873.250000  60618.000000     35.000000                 37.000000   
50%    610052.500000  60625.000000     48.000000                 49.000000   
75%    610212.750000  60639.000000     61.000000                 61.000000   
max    610544.000000  60827.000000     99.000000                 99.000000   

       Environment Score  Instruction Score  Teachers Score  \
count         513.000000         513.000000      295.000000   
mean           47.766082          48.288499       49.050847   
std            16.215584          17.417176       17.854499   
min             1.000000           1.000000        6.000000   
25%            37.000000          37.000000       36.000000   
50%            47.000000          47.000000       48.000000   
75%            58.000000          59.000000       59.000000   
max            99.000000          99.000000       99.000000   

       Parent Engagement Score  Average Student Attendance  \
count               432.000000                  565.000000   
mean                 50.231481                   92.313805   
std                   5.105125                    5.993053   
min                  37.000000                   57.900000   
25%                  47.000000                   91.900000   
50%                  50.000000                   94.400000   
75%                  53.000000                   95.600000   
max                  69.000000                   98.400000   

       Rate of Misconducts (per 100 students)   ...  \
count                               566.000000  ...   
mean                                 21.085866  ...   
std                                  27.543835  ...   
min                                   0.000000  ...   
25%                                   4.525000  ...   
50%                                  12.250000  ...   
75%                                  26.975000  ...   
max                                 251.600000  ...   

       College Enrollment (number of students)   General Services Route   \
count                                566.000000               566.000000   
mean                                 626.053004                39.084806   
std                                  448.495105                 6.165552   
min                                   21.000000                29.000000   
25%                                  342.250000                34.000000   
50%                                  504.000000                39.000000   
75%                                  790.500000                45.000000   
max                                 4368.000000                49.000000   

         RCDTS Code  X_COORDINATE  Y_COORDINATE    Latitude   Longitude  \
count  5.660000e+02  5.660000e+02  5.660000e+02  566.000000  566.000000   
mean   1.500000e+14  1.163016e+06  1.885663e+06   41.841890  -87.677323   
std    0.000000e+00  1.605930e+04  3.285688e+04    0.090361    0.058390   
min    1.500000e+14  1.118114e+06  1.817242e+06   41.653674  -87.841052   
25%    1.500000e+14  1.152166e+06  1.860248e+06   41.771866  -87.717152   
50%    1.500000e+14  1.163553e+06  1.886889e+06   41.845521  -87.675878   
75%    1.500000e+14  1.174256e+06  1.911652e+06   41.913584  -87.636059   
max    1.500000e+14  1.202811e+06  1.950960e+06   42.021064  -87.533007   

       Community Area Number        Ward  Police District  
count             566.000000  566.000000       566.000000  
mean               37.706714   22.060071        11.710247  
std                21.584135   13.645574         6.688582  
min                 1.000000    1.000000         1.000000  
25%                23.00

**Explanation:**

`.describe()` generates summary statistics (mean, median, min, max, quartiles, etc.) for all numerical columns.

**Insights Examples:**

Safety Scores average around 50, with some schools scoring close to the maximum (99). Average Student Attendance typically ranges between 90% and 100%, but a few schools report much lower values (as low as 41%), which may reflect poor attendance or missing data.



#### 7. Use .value_counts() on Categorical Columns

In [355]:
df["Elementary, Middle, or High School"].value_counts() # Counts the number of schools by type (Elementary, Middle, High)


Elementary, Middle, or High School
ES    462
HS     93
MS     11
Name: count, dtype: int64

In [357]:
df["CPS Performance Policy Status"].value_counts() # Shows how many schools fall into each CPS policy status category

CPS Performance Policy Status
Not on Probation    303
Probation           249
Not Applicable        9
NDA                   5
Name: count, dtype: int64

**Explanation:**

`.value_counts()` helps quantify how many records fall into each category. This is useful for understanding the composition of your dataset.

**Insight Examples:**

Most entries are elementary schools. There are 462 elementary schools.

Many schools are marked as “Not on Probation” or “Probation” under CPS performance status.

#### 8. Use .groupby() to Summarize Data

In [361]:
df.groupby("Elementary, Middle, or High School")["Average Student Attendance"].mean() # Groups data by school level and calculates the average student attendance in each group

Elementary, Middle, or High School
ES    94.022777
HS    83.563441
MS    94.672727
Name: Average Student Attendance, dtype: float64

**Explanation:**
This code groups the dataset by school type (Elementary, Middle, High) and calculates the average student attendance for each category.

**Insight Example:**

Elementary schools tend to have higher average attendance than high schools.

This grouping helps identify where attendance issues may be more prominent.

#### 9. Highlight 1–2 Key Takeaways or Patterns from Your Descriptive Analysis

**Takeaway 1:** 

Schools with high safety scores (above 90) tend to also have high involvement scores across families, teachers, and parents. This suggests that community engagement may be strongly linked to a school’s overall safety and environment.

**Takeaway 2:**

Elementary schools make up the largest portion of the dataset and show the highest average student attendance. This could suggest stronger student retention and engagement at earlier grade levels compared to middle or high schools.

#### 10. Suggest 2 Questions You Would Explore Further in Part 2 Using Visualization

**Question 1:** 

Do schools with higher Safety Scores also report higher Parent Engagement Scores? A scatter plot could help visualize this relationship.

**Question 2:** 

Do schools with higher teacher or parent involvement also report better performance scores (e.g., CPS Policy Status or ISAT Scores)? A boxplot comparison could help investigate this across involvement levels.

In [366]:
df.to_csv("camille_settles_cleaned.csv", index=False)  # Export the cleaned and filtered dataset 
